File: codigos/gerarCSV.m
clc
clear all


myPath = "D:\malaria\Jonathan\";
myFile = "test.json";

myFileName = strcat(myPath, myFile);

fid = fopen(myFileName); 
raw = fread(fid,inf); 
str = char(raw'); 
fclose(fid); 


dados = jsondecode(str);
tamBase = 86035;

ImageCheckSum(tamBase,1) = "";
ImagePathName(tamBase,1) = "";
ImageShapeR(tamBase,1)   = 0;
ImageShapeC(tamBase,1)   = 0;
ImageShapeChannels(tamBase,1) = 0;
ObjectsCategory(tamBase,1) = "";
ObjectsBoundingBoxMinimumR(tamBase,1) = 0;
ObjectsBoundingBoxMinimumC(tamBase,1) = 0;
ObjectsBoundingBoxMaximumR(tamBase,1) = 0;
ObjectsBoundingBoxMaximumC(tamBase,1) = 0;
InstanciaExame(tamBase,1) = 0;
Exame(tamBase,1) = 0;
Teste(tamBase,1) = false;
Treino(tamBase,1) = false;

contador = 1;
tam1 = size(dados,1);

return

% Teste
for i=1:size(dados,1)
    clear data2
    data2 = dados(i).objects;
   
    
    for j=1:size(data2,1)
        Exame(contador) = i;
        Teste(contador) = true;
        InstanciaExame(contador) = j;
        ImageCheckSum(contador)      = dados(i).image.checksum;
        ImagePathName(contador)      = dados(i).image.pathname;
        ImageShapeR(contador)        = dados(i).image.shape.r;
        ImageShapeC(contador)        = dados(i).image.shape.c;
        ImageShapeChannels(contador) = dados(i).image.shape.channels;
    
        ObjectsCategory(contador)           = data2(j).category;
        ObjectsBoundingBoxMinimumR(contador) = data2(j).bounding_box(1).minimum(1).r;
        ObjectsBoundingBoxMinimumC(contador) = data2(j).bounding_box(1).minimum(1).c;
        ObjectsBoundingBoxMaximumR(contador) = data2(j).bounding_box(1).maximum(1).r;
        ObjectsBoundingBoxMaximumC(contador) = data2(j).bounding_box(1).maximum(1).c;
        contador = contador + 1;
    end
end


myFile = "training.json";

myFileName = strcat(myPath, myFile);

fid = fopen(myFileName); 
raw = fread(fid,inf); 
str = char(raw'); 
fclose(fid); 

dados = jsondecode(str);

% Treino
for i=1:size(dados,1)
    clear data2
    data2 = dados(i).objects;
    
    for j=1:size(data2,1)
        Exame(contador) = i + tam1;
        Treino(contador) = true;
        InstanciaExame(contador) = j;
        ImageCheckSum(contador)      = dados(i).image.checksum;
        ImagePathName(contador)      = dados(i).image.pathname;
        ImageShapeR(contador)        = dados(i).image.shape.r;
        ImageShapeC(contador)        = dados(i).image.shape.c;
        ImageShapeChannels(contador) = dados(i).image.shape.channels;
    
        ObjectsCategory(contador)            = data2(j).category;
        ObjectsBoundingBoxMinimumR(contador) = data2(j).bounding_box(1).minimum(1).r;
        ObjectsBoundingBoxMinimumC(contador) = data2(j).bounding_box(1).minimum(1).c;
        ObjectsBoundingBoxMaximumR(contador) = data2(j).bounding_box(1).maximum(1).r;
        ObjectsBoundingBoxMaximumC(contador) = data2(j).bounding_box(1).maximum(1).c;
        contador = contador + 1;
    end
end

cases = sortrows(table(Exame,InstanciaExame,Teste,Treino,...
    ImageCheckSum,...
    ImagePathName,...
    ImageShapeR,...
    ImageShapeC,...
    ImageShapeChannels,...
    ObjectsBoundingBoxMinimumR, ...
    ObjectsBoundingBoxMinimumC,...
    ObjectsBoundingBoxMaximumR,...
    ObjectsBoundingBoxMaximumC,...
    ObjectsCategory), 1);

writetable(cases,'casos.csv');

File: 2_pre_processamento/gerar_csv.py
import json
import pandas as pd
from pathlib import Path
from pattern import pattern_class


path_project = Path().absolute().parent
#linux
#path_project = Path().absolute()
# Caminho para os arquivos
test_file = path_project / "1_entrada" / "test.json"
train_file = path_project / "1_entrada" / "training.json"

# Função para processar os dados JSON
def process_json(file_name, is_test=True):
    with open(file_name, 'r') as file:
        data = json.load(file)

    processed_data = []
    for i, item in enumerate(data):
        image_info = item['image']
        for j, obj in enumerate(item['objects']):
            isInfected = 1
            if(pattern_class[obj['category']] != 6 ):
                isInfected = 0 # não tá infectada
            exam_info = {
                'Exame': i + 1,
                'Teste': is_test,
                'Treino': not is_test,
                'InstanciaExame': j + 1,
                'ImageCheckSum': image_info['checksum'],
                'ImagePathName': image_info['pathname'],
                'ImageShapeR': image_info['shape']['r'],
                'ImageShapeC': image_info['shape']['c'],
                'ImageShapeChannels': image_info['shape']['channels'],
                'ObjectsCategory': isInfected,
                'ObjectsBoundingBoxMinimumR': obj['bounding_box']['minimum']['r'],
                'ObjectsBoundingBoxMinimumC': obj['bounding_box']['minimum']['c'],
                'ObjectsBoundingBoxMaximumR': obj['bounding_box']['maximum']['r'],
                'ObjectsBoundingBoxMaximumC': obj['bounding_box']['maximum']['c']
            }
            processed_data.append(exam_info)

    return processed_data

# Processar os arquivos JSON
test_data = process_json(test_file, is_test=True)
train_data = process_json(train_file, is_test=False)

# Combinar os dados de teste e treino
all_data = test_data + train_data

# Criar DataFrame do pandas e ordenar
cases_df = pd.DataFrame(all_data)
cases_df = cases_df.sort_values(by='Exame')

# Salvar como CSV
cases_df.to_csv(path_project / "2_pre_processamento" / 'casos.csv', index=False)
